org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/JavaSearchScope.java

```diff
@@ -273,9 +274,11 @@ public void add(IJavaElement element) throws JavaModelException {
 			// remember sub-cu (or sub-class file) java elements
 			if (element instanceof IMember) {
 				if (this.elements == null) {
-					this.elements = new ArrayList();
+					this.elements = new HashSet();
+					this.elementTypes = new boolean[IJavaElement.JAVA_MODULE+1];
 				}
 				this.elements.add(element);
+				this.elementTypes[element.getElementType()] = true;
```

Comment:
i see no reason for trying to cache elementTypes:

- this.elements.contains() is O(1) anyway - so no performance benefit
- wasted memory
- it is not cleaned up when removing something from this.elements

Reply:
If `elements` never contains a project or workspace why bother with the `contains`, `hashCode` and `equals` calls required to determine membership.  As for cleanup, removal is a rare case and `elementTypes` is just a guard for a more costly operation.  The array is like 12 `boolean`s so in terms of memory it is very minimal.

Reply:
Just for historical perspective, I started with the `elementType` guard and reduced search time from 45s to 25s.  The `HashSet` came next and brought it down to 20s.  There is more discussion in #474